posts/inside-rust/2023-05-03-stabilizing-async-fn-in-trait.md

```diff
@@ -0,0 +1,229 @@
+    HC: HealthCheck + Send + 'static,
+    HC::check(): Send, // <-- associated return type
+{
+    tokio::task::spawn(async move {
+        if !hc.check().await {
+            log_health_check_failure().await;
+        }
+    });
+}
+```
+
+Of course, it's kind of unfortunate that we had to rewrite from taking an `impl HealthCheck` to an explicit `HC` type parameter in order to use this notation. [RFC 2289](https://github.com/rust-lang/rfcs/pull/2289), "associated type bounds", introduced a compact notation to address this problem. That RFC is not part of this MVP, but if it were stabilized, then one could simply write:
+
+```rust
+async fn do_health_check_par(hc: impl HealthCheck<check(): Send> + Send + 'static) {
+    //                                            -------------
+    tokio::task::spawn(async move {
+        if !hc.check().await {
+            log_health_check_failure().await;
+        }
+    });
+}
+```
+
+In our [previous post][pp], we [hypothesized](https://blog.rust-lang.org/inside-rust/2022/11/17/async-fn-in-trait-nightly.html#hypothesis-this-is-uncommon) that this problem might not occur often in practice. However, our case studies found that it comes up quite frequently, and so we decided that a solution is needed. We explored a number of solutions and concluded that associated return types (ARTs) are a flexible and reasonably ergonomic building block, which makes them a great fit for an MVP.
+ 
+**Status:** Associated return types have an experimental implementation and we are currently drafting an RFC. There are several open bugs that will need to be fixed. We also found that in traits with many methods, ARTs become verbose, and in the future may consider more concise syntax (see below).
+
+[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=2066934a05cb9eafc0b47af7bdf8c57f)
+
+## MVP Part 3: "impl trait in traits" (return position)
+
+In Rust an async function is "syntactic sugar" for a function that returns an `impl Future`, and async functions in traits are no exception. As part of the MVP, we plan to stabilize the use of `-> impl Trait` notation in traits and trait impls.
+
+Impl trait in traits has all kinds of uses, but one common one for async programming is to avoid capturing all of the function arguments by doing some amount of sync work and then returning a future for the rest. For example, this `LaunchService` trait declares a `launch` function that does not capture `self` (similar to the existing Tower [`Service`] trait):
+
+[`Service`]: https://docs.rs/tower/latest/tower/trait.Service.html
+
+```rust
+trait LaunchService {
+    fn launch(
+        &mut self, 
+        request: Request,
+    ) -> impl Future<Output = u32>;
+    //   -------------------------
+    //   Does not capture `self` as it does
+    //   not include a `+ '_`.
+}
+```
+
+Since `async fn` is sugar for a regular function returning `impl Future`, these two syntactic forms will work interchangeably.
```

Comment:
If I remember right, the original AFIT RFC said specifically we wouldn't allow mixing and matching at first, right? Do we want to call out the design change?

Reply:
It's called out in the RPITIT RFC that's up right now. I don't think we need to highlight it in this post so much, but would be fine with adding a footnote or something like that.

Reply:
Sounds good. Yeah, the RFC is the right place to mention that. It's probably too small of a detail to call out in this blog post.